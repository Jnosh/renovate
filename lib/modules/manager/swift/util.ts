import { isVersion as isSemverVersion } from '../../../modules/versioning/semver';
import { regEx } from '../../../util/regex';

// Ignore (truncate) any pre-release component for now since SPM is currently documented to ignore them as well
const SWIFT_TOOLS_VERSION_REGEX = regEx(
  /^\s*\/\/\s*swift-tools-version:\s*(?<version>\d+\.\d+(?:\.\d+)?)/
);

/**
 * Extracts the Swift tools version from the given package file contents as a semver version.
 *
 * Tools version specification:
 * https://github.com/apple/swift-package-manager/blob/main/Documentation/Usage.md#swift-tools-version-specification
 */
export function extractSwiftToolsVersion(
  packageFileContent: string
): string | null {
  const toolsVersionMatch = SWIFT_TOOLS_VERSION_REGEX.exec(packageFileContent);

  if (toolsVersionMatch?.groups) {
    const version = toolsVersionMatch.groups.version;

    // The version is either '<major>.<minor>.<patch>' or '<major>.<minor>'
    // Ensure we always return a semver compatible version.
    return isSemverVersion(version) ? version : `${version}.0`;
  }

  return null;
}
