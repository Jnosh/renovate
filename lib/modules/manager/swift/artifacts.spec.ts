import { join } from 'upath';
import { envMock, mockExecAll } from '../../../../test/exec-util';
import { env, fs, git } from '../../../../test/util';
import { GlobalConfig } from '../../../config/global';
import type { RepoGlobalConfig } from '../../../config/types';
import * as docker from '../../../util/exec/docker';
import type { UpdateArtifactsConfig } from '../types';
import * as swift from '.';

jest.mock('../../../util/exec/env');
jest.mock('../../../util/git');
jest.mock('../../../util/http');
jest.mock('../../../util/fs');

const config: UpdateArtifactsConfig = {};

const adminConfig: RepoGlobalConfig = {
  // `join` fixes Windows CI
  localDir: join('/tmp/github/some/repo'),
  cacheDir: join('/tmp/cache'),
  containerbaseDir: join('/tmp/cache/containerbase'),
};

describe('modules/manager/swift/artifacts', () => {
  beforeEach(() => {
    jest.resetAllMocks();
    jest.resetModules();

    env.getChildProcessEnv.mockReturnValue(envMock.basic);
    GlobalConfig.set(adminConfig);
    docker.resetPrefetchedImages();
  });

  afterEach(() => {
    GlobalConfig.reset();
  });

  it('returns null if no Package.resolved file found', async () => {
    fs.getSiblingFileName.mockReturnValueOnce('Package.resolved');
    fs.localPathExists.mockResolvedValueOnce(false);
    const updatedDeps = [
      {
        depName: 'dep1',
      },
    ];
    expect(
      await swift.updateArtifacts({
        packageFileName: 'Package.swift',
        updatedDeps,
        newPackageFileContent: '',
        config,
      })
    ).toBeNull();
  });

  it('returns null if updatedDeps is empty', async () => {
    expect(
      await swift.updateArtifacts({
        packageFileName: 'Package.swift',
        updatedDeps: [],
        newPackageFileContent: '',
        config,
      })
    ).toBeNull();
  });

  it('returns null if unchanged', async () => {
    fs.getSiblingFileName.mockReturnValueOnce('Package.resolved');
    fs.localPathExists.mockResolvedValueOnce(true);
    fs.readLocalFile.mockResolvedValueOnce('Current Package.resolved');
    fs.getParentDir.mockReturnValueOnce('');
    const execSnapshots = mockExecAll();
    fs.readLocalFile.mockResolvedValueOnce('Current Package.resolved');

    const updatedDeps = [
      {
        depName: 'dep1',
      },
    ];
    expect(
      await swift.updateArtifacts({
        packageFileName: 'Package.swift',
        updatedDeps,
        newPackageFileContent: '',
        config,
      })
    ).toBeNull();
    expect(execSnapshots).toMatchSnapshot();
  });

  it('returns updated Package.resolved', async () => {
    fs.getSiblingFileName.mockReturnValueOnce('Package.resolved');
    fs.localPathExists.mockResolvedValueOnce(true);
    git.getFile.mockResolvedValueOnce('Old Package.resolved');
    fs.getParentDir.mockReturnValueOnce('');
    const execSnapshots = mockExecAll();
    fs.readLocalFile.mockResolvedValueOnce('New Package.resolved');
    const updatedDeps = [
      {
        depName: 'dep1',
      },
    ];
    expect(
      await swift.updateArtifacts({
        packageFileName: 'Package.swift',
        updatedDeps,
        newPackageFileContent: '{}',
        config,
      })
    ).not.toBeNull();
    expect(execSnapshots).toMatchSnapshot();
  });

  it('handles package in subfolder', async () => {
    fs.getSiblingFileName.mockReturnValueOnce('sub/path/Package.resolved');
    fs.localPathExists.mockResolvedValueOnce(true);
    git.getFile.mockResolvedValueOnce('Old Package.resolved');
    fs.getParentDir.mockReturnValueOnce('sub/path/');
    const execSnapshots = mockExecAll();
    fs.readLocalFile.mockResolvedValueOnce('New Package.resolved');
    const updatedDeps = [
      {
        depName: 'dep1',
      },
    ];
    expect(
      await swift.updateArtifacts({
        packageFileName: 'sub/path/Package.swift',
        updatedDeps,
        newPackageFileContent: '{}',
        config,
      })
    ).not.toBeNull();
    expect(execSnapshots).toMatchSnapshot();
  });

  it('returns updated Package.resolved for lockfile maintenance', async () => {
    fs.getSiblingFileName.mockReturnValueOnce('Package.resolved');
    fs.localPathExists.mockResolvedValueOnce(true);
    git.getFile.mockResolvedValueOnce('Old Package.resolved');
    fs.getParentDir.mockReturnValueOnce('');
    const execSnapshots = mockExecAll();
    fs.readLocalFile.mockResolvedValueOnce('New Package.resolved');
    expect(
      await swift.updateArtifacts({
        packageFileName: 'Package.swift',
        updatedDeps: [],
        newPackageFileContent: '{}',
        config: { ...config, updateType: 'lockFileMaintenance' },
      })
    ).not.toBeNull();
    expect(execSnapshots).toMatchSnapshot();
  });

  it('returns updated Package.resolved with docker', async () => {
    fs.getSiblingFileName.mockReturnValueOnce('Package.resolved');
    fs.localPathExists.mockResolvedValueOnce(true);
    GlobalConfig.set({ ...adminConfig, binarySource: 'docker' });
    git.getFile.mockResolvedValueOnce('Old Package.resolved');
    fs.getParentDir.mockReturnValueOnce('');
    const execSnapshots = mockExecAll();
    fs.readLocalFile.mockResolvedValueOnce('New Package.resolved');
    const updatedDeps = [
      {
        depName: 'dep1',
      },
    ];
    expect(
      await swift.updateArtifacts({
        packageFileName: 'Package.swift',
        updatedDeps,
        newPackageFileContent: '{}',
        config,
      })
    ).not.toBeNull();
    expect(execSnapshots).toMatchSnapshot();
  });

  it('catches errors', async () => {
    fs.getSiblingFileName.mockReturnValueOnce('Package.resolved');
    fs.localPathExists.mockResolvedValueOnce(true);
    fs.readLocalFile.mockResolvedValueOnce('Current Package.resolved');
    fs.writeLocalFile.mockImplementationOnce(() => {
      throw new Error('not found');
    });
    const updatedDeps = [
      {
        depName: 'dep1',
      },
    ];
    expect(
      await swift.updateArtifacts({
        packageFileName: 'Package.swift',
        updatedDeps,
        newPackageFileContent: '{}',
        config,
      })
    ).toEqual([
      { artifactError: { lockFile: 'Package.resolved', stderr: 'not found' } },
    ]);
  });
});
