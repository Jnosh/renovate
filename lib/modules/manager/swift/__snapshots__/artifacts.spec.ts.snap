// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`modules/manager/swift/artifacts handles package in subfolder 1`] = `
[
  {
    "cmd": "swift package resolve --package-path sub/path/",
    "options": {
      "cwd": "/tmp/github/some/repo",
      "encoding": "utf-8",
      "env": {
        "HOME": "/home/user",
        "HTTPS_PROXY": "https://example.com",
        "HTTP_PROXY": "http://example.com",
        "LANG": "en_US.UTF-8",
        "LC_ALL": "en_US",
        "NO_PROXY": "localhost",
        "PATH": "/tmp/path",
      },
      "maxBuffer": 10485760,
      "timeout": 900000,
    },
  },
]
`;

exports[`modules/manager/swift/artifacts returns null if unchanged 1`] = `
[
  {
    "cmd": "swift package resolve --package-path ''",
    "options": {
      "cwd": "/tmp/github/some/repo",
      "encoding": "utf-8",
      "env": {
        "HOME": "/home/user",
        "HTTPS_PROXY": "https://example.com",
        "HTTP_PROXY": "http://example.com",
        "LANG": "en_US.UTF-8",
        "LC_ALL": "en_US",
        "NO_PROXY": "localhost",
        "PATH": "/tmp/path",
      },
      "maxBuffer": 10485760,
      "timeout": 900000,
    },
  },
]
`;

exports[`modules/manager/swift/artifacts returns updated Package.resolved 1`] = `
[
  {
    "cmd": "swift package resolve --package-path ''",
    "options": {
      "cwd": "/tmp/github/some/repo",
      "encoding": "utf-8",
      "env": {
        "HOME": "/home/user",
        "HTTPS_PROXY": "https://example.com",
        "HTTP_PROXY": "http://example.com",
        "LANG": "en_US.UTF-8",
        "LC_ALL": "en_US",
        "NO_PROXY": "localhost",
        "PATH": "/tmp/path",
      },
      "maxBuffer": 10485760,
      "timeout": 900000,
    },
  },
]
`;

exports[`modules/manager/swift/artifacts returns updated Package.resolved for lockfile maintenance 1`] = `
[
  {
    "cmd": "swift package resolve --package-path ''",
    "options": {
      "cwd": "/tmp/github/some/repo",
      "encoding": "utf-8",
      "env": {
        "HOME": "/home/user",
        "HTTPS_PROXY": "https://example.com",
        "HTTP_PROXY": "http://example.com",
        "LANG": "en_US.UTF-8",
        "LC_ALL": "en_US",
        "NO_PROXY": "localhost",
        "PATH": "/tmp/path",
      },
      "maxBuffer": 10485760,
      "timeout": 900000,
    },
  },
]
`;

exports[`modules/manager/swift/artifacts returns updated Package.resolved with docker 1`] = `
[
  {
    "cmd": "docker pull renovate/swift",
    "options": {
      "encoding": "utf-8",
    },
  },
  {
    "cmd": "docker ps --filter name=renovate_swift -aq",
    "options": {
      "encoding": "utf-8",
    },
  },
  {
    "cmd": "docker run --rm --name=renovate_swift --label=renovate_child -v "/tmp/github/some/repo":"/tmp/github/some/repo" -v "/tmp/cache":"/tmp/cache" -e BUILDPACK_CACHE_DIR -w "/tmp/github/some/repo" renovate/swift bash -l -c "swift package resolve --package-path ''"",
    "options": {
      "cwd": "/tmp/github/some/repo",
      "encoding": "utf-8",
      "env": {
        "BUILDPACK_CACHE_DIR": "/tmp/cache/containerbase",
        "HOME": "/home/user",
        "HTTPS_PROXY": "https://example.com",
        "HTTP_PROXY": "http://example.com",
        "LANG": "en_US.UTF-8",
        "LC_ALL": "en_US",
        "NO_PROXY": "localhost",
        "PATH": "/tmp/path",
      },
      "maxBuffer": 10485760,
      "timeout": 900000,
    },
  },
]
`;
